keps/sig-cluster-lifecycle/kubeadm/20190722-Advanced-configurations-with-kubeadm-(Kustomize).md

```diff
@@ -0,0 +1,292 @@
+
+## Proposal
+
+### User Stories 
+
+#### Story 1
+As a cluster administrator, I want to set kubelet eviction policy for a joining node
+according to the node hardware features.
+
+#### Story 2
+As a cluster administrator, I want to set timeouts for the kube-apiserver liveness
+probes for edge clusters.
+
+#### Story 3
+As a cluster administrator, I want to upgrade my cluster preserving all the
+“advanced configuration” already in place
+
+### Implementation Details
+
+This proposal explores as a first option for implementing Kubeadm 
+“advance configrations“ the usage of Kustomize; please refer to
+[Declarative application management in Kubernetes](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/architecture/declarative-application-management.md)
+and [Kustomize KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cli/0008-kustomize.md)
+for background information about Kustomize.
+
+Support for "advanced configrations“/kustomize in kubeadm will be available under
+a new feature-flag named `kustomize`, initially alpha and disabled by default.
+
+#### Kustomize integration with kubeadm
+
+By adopting Kustomize in the context of Kubeadm, this proposal assumes to:
+
+- Let kubeadm generate static pod manifests or kubelet configuration as usual.
+- Use kubeadm generated artifacts as a starting point for applying patches
+  containing “advanced configurations”.
+
+This has some implications:
+
+1. The higher-level tools/users have to express “advanced configurations” using
+   one of the two alternative techniques supported by Kustomize - the [strategic
+   merge patch](https://github.com/kubernetes-sigs/kustomize/blob/master/docs/glossary.md#patchstrategicmerge)
+   and the [JSON patch](https://github.com/kubernetes-sigs/kustomize/blob/master/docs/glossary.md#patchjson6902) -.
+2. The higher-level tools/users have to provide patches before running kubeadm;
+   this point is going to be further discussed in the following paragraphs.
+3. Kubeadm is responsible for coordinating the execution of Kustomize within the
+   init/join/upgrade workflows
+4. as a consequence of the previous point, higher-level tools/users are not
+   requested to take care of defining `kustomization.yaml` files nor to define
+   a local folder structure.
+
+Additionally, in order to simplify the first implementation of this KEP, this 
+proposal is going to assume that Kustomize patches for kubeadm are always defined
+specifically for the node where kubeadm is going to be executed.
+
+This point could be reconsidered in the future, by e.g. introducing cluster-wide
+patches and/or patches for a subset of nodes.
+
+#### Providing and storing Kustomize patches to kubeadm
+
+Before kubeadm init, Kustomize patches should be eventually provided to kubeadm
+by higher-level tools/users; patches should be provided in a well know location
+like e.g. `/etc/kubernetes/kustomize` or in a custom location specified with a CLI flag. 
```

Comment:
as noted in my discussions with Fabrizio, `/etc/kubernetes/kustomize` is tricky because it is a location managed by kubeadm, thus currently  `reset` would clean this location. then again the same goes for user certs copied to `../pki`.

Reply:
I cannot help but think that there are two different set of patches we want to apply. Unless I'm missing some use case I think it would be reasonable for control planes to share the same configuration, so the set of patches to be applied on control plane components are saved cluster wide in a `ConfigMap`. Node specific patches could follow this approach of a well known location though, until we maybe evolve in something cluster-wide as well as described in the document.

Reply:
> /etc/kubernetes/kustomize is tricky

Removed. The user chooses its preferred location for patches

> I cannot help but think that there are two different set of patches we want to apply. 

Agreed but intentionally ruled out of this first iteration for sake of simplicity
see _"Additionally, in order to simplify the first implementation of this KEP, this 
proposal is going to assume that Kustomize patches for kubeadm are always defined
specifically for the node where kubeadm is going to be executed.

This point could be reconsidered in the future, by e.g. introducing cluster-wide
patches and/or patches for a subset of nodes."_